README.md
Bowline
-------

An involuntary API for Supermechanical's [Twine][0].

[![Build Status](https://travis-ci.org/banterability/bowline.svg?branch=master)](https://travis-ci.org/banterability/bowline)


## Installation

```bash
npm install bowline
```


## Usage

Key        | Value
---        | -----
`email`    | Your email address for [twine.cc][1].
`password` | Your password for [twine.cc][1].
`deviceId` | Your Twine's unique ID. You can find this in the URL of your Twine's "Rules" page – see [this (hopefully) helpful diagram][2].

[0]: http://supermechanical.com/twine/
[1]: https://twine.cc/login
[2]: http://cl.ly/TeNO/Screen_Shot_2014-01-30_at_12.40.48_PM.jpg

```javascript
var Bowline = require('bowline');

var client = new Bowline({
  email: 'twine@example.net',
  password: 'sekret',
  deviceId: '00ff00ff00ff00ff'
});

client.fetch(function(err, response){
  // do something
});
```


## Response Structure

```javascript
{
  meta: {
    battery: 'weak',
    wifiSignal: 'full'
  },
  time: {
    age: 32.436,
    timestamp: 1418069760
  },
  values: {
    batteryVoltage: 2942324,
    firmwareVersion: '2.0.2'
    isVibrating: false,
    orientation: 'top'
    temperature: 68,
    updateMode: 'normal',
    vibration: 0
  }
}
```

### Meta

Key          | Sample Value | Notes
---          | ------------ | -----
`battery`    | `'1/3'`      | Possible values: `plugged in`, `full`, `2/3`, `1/3`, `weak`, `unknown`
`wifiSignal` | `'full'`     | Possible values: `full`, `2/3`, `1/3`, `weak`, `unknown`

### Time

Key         | Sample Value | Notes
---         | ------------ | -----
`age`       | `32.436`     | Seconds since data updated
`timestamp` | `1418069760` | Unix timestamp of data update time

### Values

Key               | Sample Value | Notes
---               | ------------ | -----
`firmwareVersion` | `'2.0.2'`    |
`temperature`     | `70`         | in degress Fahrenheit
`vibration`       | `0`          | in milli-Gs
`orientation`     | `'top'`      | Possible values: `top`, `bottom`, `left`, `right`, `front`, `back`
`isVibrating`     | `false`      | Possible values: `true`, `false`
`batteryVoltage`  | `2.68`       | in volts
`updateMode`      | `'normal'`   | Possible values: `normal`, `fast`


## Development

Command      | Description
-------      | ------------
`make build` | Compile from Coffeescript
`make test`  | Compile and run tests


CHANGELOG.md
2.0.0
-----
* [BREAKING] Battery voltage now returns volts instead of microvolts
* Response includes descriptions of battery & Wi-Fi strength as `meta`

1.0.0
-----
* [BREAKING] Values are now a key in the response, not at the top level
* Response includes time information

0.2.0
-----
* Only authenticate once per session

0.1.0
-----
* Initial prerelease


lib/presenter.js
// Generated by CoffeeScript 1.8.0
var VALUE_MAP, makeValuesUseful, presentValues;

VALUE_MAP = {
  "00": 'firmwareVersion',
  "01": 'temperature',
  "02": 'vibration',
  "03": 'orientation',
  "04": 'isVibrating',
  "05": 'batteryVoltage',
  "07": 'updateMode'
};

module.exports = function(response) {
  return {
    values: presentValues(response)
  };
};

presentValues = function(response) {
  var friendlyName, identifier, key, output, tuple, value, _i, _len, _ref;
  output = {};
  _ref = response.values;
  for (_i = 0, _len = _ref.length; _i < _len; _i++) {
    tuple = _ref[_i];
    key = tuple[0], value = tuple[1];
    identifier = key.substr(-2, 2);
    if (friendlyName = VALUE_MAP[identifier]) {
      output[friendlyName] = value;
    }
  }
  return makeValuesUseful(output);
};

makeValuesUseful = function(values) {
  values.temperature = values.temperature / 100;
  values.isVibrating = values.isVibrating === 0 ? false : true;
  values.updateMode = values.updateMode === 0 ? 'normal' : 'fast';
  return values;
};


lib/presenters/meta.js
// Generated by CoffeeScript 1.8.0
var presentBatteryStrength, presentMetadata, presentWiFiStrength;

presentMetadata = function(response, _arg) {
  var batteryVoltage;
  batteryVoltage = _arg.batteryVoltage;
  return {
    battery: presentBatteryStrength(batteryVoltage),
    wifiSignal: presentWiFiStrength(response)
  };
};

presentBatteryStrength = function(voltage) {
  if (voltage) {
    if (voltage < 1.3) {
      return 'plugged in';
    }
    switch (false) {
      case !(voltage > 2.75):
        return 'full';
      case !(voltage > 2.6):
        return '2/3';
      case !(voltage > 2.45):
        return '1/3';
      default:
        return 'weak';
    }
  } else {
    return 'unknown';
  }
};

presentWiFiStrength = function(response) {
  var strength;
  if (strength = response.rssi) {
    switch (false) {
      case !(strength > 203):
        return 'full';
      case !(strength > 195):
        return '2/3';
      case !(strength > 185):
        return '1/3';
      default:
        return 'weak';
    }
  } else {
    return 'unknown';
  }
};

module.exports = presentMetadata;


lib/presenters/values.js
// Generated by CoffeeScript 1.8.0
var VALUE_MAP, makeValuesUseful, presentValues;

VALUE_MAP = {
  "00": 'firmwareVersion',
  "01": 'temperature',
  "02": 'vibration',
  "03": 'orientation',
  "04": 'isVibrating',
  "05": 'batteryVoltage',
  "07": 'updateMode'
};

presentValues = function(response) {
  var friendlyName, identifier, key, output, tuple, value, _i, _len, _ref;
  output = {};
  _ref = response.values;
  for (_i = 0, _len = _ref.length; _i < _len; _i++) {
    tuple = _ref[_i];
    key = tuple[0], value = tuple[1];
    identifier = key.substr(-2, 2);
    if (friendlyName = VALUE_MAP[identifier]) {
      output[friendlyName] = value;
    }
  }
  return makeValuesUseful(output);
};

makeValuesUseful = function(values) {
  values.temperature = values.temperature / 100;
  values.isVibrating = values.isVibrating === 0 ? false : true;
  values.batteryVoltage = Math.round(values.batteryVoltage / 10000) / 100;
  values.updateMode = values.updateMode === 0 ? 'normal' : 'fast';
  return values;
};

module.exports = presentValues;
